dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/TrackerBundleParamsConverter.java

```diff
@@ -84,174 +82,251 @@
-        for ( Enrollment enrollment : bundle.getTrackedEntities().stream()
-            .flatMap( l -> l.getEnrollments().stream() ).collect( Collectors.toList() ) )
+        // Extract all relationships
+        for ( Event event : eventHashMap.values() )
         {
-            // collect all events from enrollments and add them to the flattened events collection
-            events.addAll( enrollment.getEvents().stream().map( e -> addParent( e, enrollment.getEnrollment() ) )
-                .collect( Collectors.toList() ) );
-
-            // Extract relationships from enrollment
-            relationships.addAll( enrollment.getRelationships() );
-
-            // Extract relationships from events
-            relationships
-                .addAll( events.stream().flatMap( l -> l.getRelationships().stream() ).collect( Collectors.toSet() ) );
-
-            // remove events and relationships from enrollment
-            enrollment.setEvents( Collections.emptyList() );
-            enrollment.setRelationships( Collections.emptyList() );
-            // remove relationships from events
-            events.forEach( e -> e.setRelationships( Collections.emptyList() ) );
+            extractRelationships( event )
+                .forEach( relationship -> relationshipHashMap.put( relationship.getRelationship(), relationship ) );
         }
 
-        for ( TrackedEntity trackedEntity : bundle.getTrackedEntities() )
-        {
-            enrollments.addAll( trackedEntity.getEnrollments().stream()
-                .map( e -> addParent( e, trackedEntity.getTrackedEntity() ) ).collect( Collectors.toList() ) );
-
-            // Extract relationships from Tracked Entity
-            relationships.addAll( trackedEntity.getRelationships() );
+        // Set UID for all relationships
+        dataBundle.getRelationships().stream()
+            .peek( this::updateRelationshipReferences )
+            .forEach( relationship -> relationshipHashMap.put( relationship.getRelationship(), relationship ) );
+
+        enrollmentHashMap.values()
+            .forEach( enrollment -> enrollment
+                .setNotes(
+                    enrollment.getNotes().stream()
+                        .filter( note -> !StringUtils.isEmpty( note.getValue() ) )
+                        .collect( Collectors.toList() )
+                ) );
+
+        eventHashMap.values()
+            .forEach(
+                event -> event.setNotes( event.getNotes().stream()
+                    .filter( note -> !StringUtils.isEmpty( note.getValue() ) )
+                    .collect( Collectors.toList() )
+                ) );
+
+        return TrackerBundleParams.builder()
+            .trackedEntities( new ArrayList<>( trackedEntityMap.values() ) )
+            .enrollments( new ArrayList<>( enrollmentHashMap.values() ) )
+            .events( new ArrayList<>( eventHashMap.values() ) )
+            .relationships( relationshipHashMap.values().stream()
```

Comment:
If 2 identical relationships without UID are in the payload, after updating the reference I think that the distinct will not recognise them as equals.

Reply:
You are right :) This is solved by Luciano's changes anyway, so I will just remove it!